extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/DifferenceWithinRangeScrutinizer.java

```diff
@@ -0,0 +1,53 @@
+package org.openrefine.wikidata.qa.scrutinizers;
+
+import org.openrefine.wikidata.qa.QAWarning;
+import org.openrefine.wikidata.updates.ItemUpdate;
+import org.wikidata.wdtk.datamodel.interfaces.*;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DifferenceWithinRangeScrutinizer extends EditScrutinizer {
+
+    public static final String type = "difference-of-the-properties-is-not-within-the-specified-range";
+
+    @Override
+    public void scrutinize(ItemUpdate update) {
+        Map<PropertyIdValue, Value> propertyIdValueValueMap = new HashMap<>();
+        for (Statement statement : update.getAddedStatements()){
+            PropertyIdValue pid = statement.getClaim().getMainSnak().getPropertyId();
+            Value value = statement.getClaim().getMainSnak().getValue();
+            propertyIdValueValueMap.put(pid, value);
+        }
+
+        for(PropertyIdValue propertyId : propertyIdValueValueMap.keySet()){
+            if (_fetcher.hasDiffWithinRange(propertyId)){
+                PropertyIdValue lowerPropertyId = _fetcher.getLowerPropertyId(propertyId);
+                QuantityValue minRangeValue = _fetcher.getMinimumValue(propertyId);
+                QuantityValue maxRangeValue = _fetcher.getMaximumValue(propertyId);
+
+                if (propertyIdValueValueMap.containsKey(lowerPropertyId)){
+                    Value startingValue = propertyIdValueValueMap.get(lowerPropertyId);
+                    Value endingValue = propertyIdValueValueMap.get(propertyId);
+
+                    if (startingValue instanceof TimeValue && endingValue instanceof TimeValue){
+                        TimeValue lowerDate = (TimeValue)startingValue;
+                        TimeValue upperDate = (TimeValue)endingValue;
+
+                        if (upperDate.getYear() - lowerDate.getYear() < minRangeValue.getNumericValue().longValue()
+                            || upperDate.getYear() == lowerDate.getYear() && upperDate.getMonth() < lowerDate.getMonth()
+                            || upperDate.getYear() == lowerDate.getYear() && upperDate.getMonth() == lowerDate.getMonth() && upperDate.getDay() < lowerDate.getDay()
+                            || upperDate.getYear() - lowerDate.getYear() > maxRangeValue.getNumericValue().longValue()){
```

Comment:
I think this would be cleaner if we first compute the difference between the two values, in some unit, and then compare it to the bounds supplied in the constraint.

Also, it is not clear to me how this works in the Wikibase implementation, but I think the constraint system automatically converts units: differences can be specified in any unit, as far as I can tell.

See the official docs:

> WikibaseQualityConstraints normalizes the range endpoints before checking this constraint; for example, the intervals [ 0 days , 14 days] and [ 0 weeks , 2 weeks] are equivalent.

https://www.wikidata.org/wiki/Help:Property_constraints_portal/Diff_within_range

Reply:
If I'll even calculate the difference that would be in terms of year-month-day and the range is provided in years. So, 1st thing that comes in my mind is it might increase the complexity of the code to understand the working and 2nd thing is the comparison with range. 

for instance: here it's quite simple to even check for same death and birth of year but an invalid month and day.  In case of difference, it will be like 0 years -2 month something which doesn't sound good to me. I also checked wikibase implementation before implementing this. They only checked for the year difference which I guess not so appropriate in cases like:
D.O.B : 2020-10-09
D.O.D: 2020-02-02
(ps: I'm not too good in PHP maybe I have overlooked any condition)
Do you want to have that implementation, I can proceed with that. No issues :)

Reply:
Oh indeed, they only support years on their end too. That's not really what the docs say!

Still, I would prefer to compute the difference in years first, and then check how it compares to the bounds, which is what they do:
https://github.com/wikimedia/mediawiki-extensions-WikibaseQualityConstraints/blob/master/src/ConstraintCheck/Checker/DiffWithinRangeChecker.php#L166